passes/cmds/exec.cc

```diff
@@ -0,0 +1,190 @@
+			std::string str;
+			YS_REGEX_TYPE re;
+
+			expect_stdout_elem() : matched(false), polarity(true), str(), re(){};
+		};
+		std::vector<expect_stdout_elem> expect_stdout;
+
+		if(args.size() == 0)
+			log_cmd_error("No command provided.\n");
+
+		for(size_t argidx = 1; argidx < args.size(); ++argidx) {
+			if (flag_cmd) {
+				cmd += args[argidx] + (argidx != (args.size() - 1)? " " : "");
+			} else {
+				if (args[argidx] == "--")
+					flag_cmd = true;
+				else if (args[argidx] == "-q")
+					flag_quiet = true;
+				else if (args[argidx] == "-expect-return") {
+					flag_expect_return = true;
+					++argidx;
+					if (argidx >= args.size())
+						log_cmd_error("No expected return value specified.\n");
+
+					expect_return_value = atoi(args[argidx].c_str());
+				} else if (args[argidx] == "-expect-stdout") {
+					flag_expect_stdout = true;
+					++argidx;
+					if (argidx >= args.size())
+						log_cmd_error("No expected regular expression specified.\n");
+
+					try{
+						expect_stdout_elem x;
+						x.str = args[argidx];
+						x.re = YS_REGEX_COMPILE(args[argidx]);
+						expect_stdout.push_back(x);
+					} catch (const YS_REGEX_NS::regex_error& e) {
+						log_cmd_error("Error in regex expression '%s' !\n", args[argidx].c_str());
+					}
+				} else if (args[argidx] == "-not-expect-stdout") {
+					flag_expect_stdout = true;
+					++argidx;
+					if (argidx >= args.size())
+						log_cmd_error("No expected regular expression specified.\n");
+
+					try{
+						expect_stdout_elem x;
+						x.str = args[argidx];
+						x.re = YS_REGEX_COMPILE(args[argidx]);
+						x.polarity = false;
+						expect_stdout.push_back(x);
+					} catch (const YS_REGEX_NS::regex_error& e) {
+						log_cmd_error("Error in regex expression '%s' !\n", args[argidx].c_str());
+					}
+
+				} else
+					log_cmd_error("Unknown option \"%s\" or \"--\" doesn\'t precede command.", args[argidx].c_str());
+			}
+		}
+
+		log_header(design, "Executing command \"%s\".\n", cmd.c_str());
+		log_push();
+
+		fflush(stdout);
+		bool keep_reading = true;
+		auto *f = popen(cmd.c_str(), "r");
+		if (f == nullptr)
+			log_cmd_error("errno %d after popen() returned NULL.\n", errno);
+		while (keep_reading) {
+			keep_reading = (fgets(buf, sizeof(buf), f) != nullptr);
+			linebuf += buf;
+			memset(buf, 0, sizeof(buf));
+
+			auto pos = linebuf.find('\n');
+			while (pos != std::string::npos) {
+				std::string line = linebuf.substr(0, pos);
+				linebuf.erase(0, pos + 1);
+				if (!flag_quiet)
+					log("%s\n", line.c_str());
+
+				if (flag_expect_stdout)
+					for(auto &x : expect_stdout)
+						if (YS_REGEX_NS::regex_search(line, x.re))
+							x.matched = true;
+
+				pos = linebuf.find('\n');
+			}
+		}
+		int status = pclose(f);
+		int retval = -1;
+
+		if(WIFEXITED(status)) {
+		    retval = WEXITSTATUS(status);
+		}
+		else if(WIFSIGNALED(status)) {
+		    retval = WTERMSIG(status);
+		}
+		else if(WIFSTOPPED(status)) {
+		    retval = WSTOPSIG(status);
+		}
+
```

Comment:
Since this is not supported on all platforms, I would suggest something like https://github.com/YosysHQ/yosys/blob/master/kernel/yosys.cc#L363-L369

Reply:
Should be taken care of by defining these macros manually for Win32 platforms.